src/nfa/nfa.cc

```diff
@@ -1167,14 +1180,32 @@ Nfa Mata::Nfa::reduce(const Nfa &aut, StateToStateMap *state_map, const StringMa
     if ("simulation" == algorithm) {
         if (state_map == nullptr) {
             std::unordered_map<State,State> tmp_state_map;
-            return reduce_size_by_simulation(aut, tmp_state_map);
+            result = reduce_size_by_simulation(aut, tmp_state_map);
         } else {
-            return reduce_size_by_simulation(aut, *state_map);
+            state_map->clear();
+            result = reduce_size_by_simulation(aut, *state_map);
         }
     } else {
         throw std::runtime_error(std::to_string(__func__) +
                                  " received an unknown value of the \"algorithm\" key: " + algorithm);
     }
+
+    if (trim_result) {
+        std::unordered_map<State,State> trimmed_state_map;
+        result.trim(&trimmed_state_map);
+        if (state_map) {
+            StateToStateMap result_state_map{ *state_map };
+            for (const auto& reduced_mapping : *state_map) {
+                const auto trimmed_mapping{ trimmed_state_map.find(reduced_mapping.second) };
```

Comment:
Shouldn't we look for `trimmed_mapping` that maps to `reduced_mapping.second` and not from?

Reply:
My bad, it is correct